### PR TITLE
Merge clangd's `inactiveRegions` with existing diagnostics

### LINKED ISSUE #13089: Dimming #if / #ifndef / ifdef macro blocks in C/C++ (inactiveRegions)
### Summary

Support dimming inactive code regions via inactiveRegions provided by clangd LSP.

### Description

When #if / #ifndef / #ifdef will not be compiled, the code block inside it will be dimmed.

### Screenshots

image

### PR DESCRIPTION
**Pull request by naim94a**
Closes: https://github.com/zed-industries/zed/issues/13089

This PR attempts to resolve the issues discussed in my previous PR #26146.

TODOs:
- [X] merge`inactiveRegions` with existing diagnostics.
- [x] save `inactiveRegions` aside and merge in the `CLspAdapter`s `process_diagnostics` to avoid blinking of `inactiveRegions`.

Release Notes:

- Fixed: `inactiveRegions` doesn't replace existing diagnostics anymore

### CODE REVIEW

File: crates/project/src/lsp_store.rs [hunk 1]
```diff
@@ -5812,6 +5823,30 @@ impl LspStore {
...
+                        .map(|v| {
+                            let start = Unclipped(v.range.start.to_point_utf16(&snapshot));
+                            let end = Unclipped(v.range.end.to_point_utf16(&snapshot));
+                            let range = start..end;
+                            DiagnosticEntry {
+                                range,
+                                diagnostic: v.diagnostic.clone(),
+                            }
+                        })
+                        .collect::<Vec<_>>()
```
Comment (SomeoneToIgnore) [inline crates/project/src/lsp_store.rs:5844]:
```suggestion
```

Can also do `.get_diagnostics(server_id).into_iter().flat_map(...` and avoid any extra nesting and `if let Some`.

**Review by SomeoneToIgnore — COMMENTED** [review]
Looks good and quite simple, thank you for the first step.

Curious about inactive regions' data invalidation: would a server send empty inactive regions, or how are we planning to understand the data is obsolete?

**Comment by naim94a** [comment]
Unlike diagnostics, clangd doesn't send an empty array of diagnostic on document close. I think that we need to treat `inactiveRegion` as an extension of diagnostics. 
The only issue now is getting the existing diagnostics and only retaining `inactiveRegion`s within `languages/src/c.rs` until the next `inactiveRegion` notification arrives - this should prevent blinking.

**Comment by naim94a** [comment]
clangd does not send an empty `inactiveRegions` on document close. In a previous comment (which I deleted) I thought we could just remove everything if diagnostics are empty, but diagnostics could also be empty if the document is well written... I'm not sure how to handle `inactiveRegions` invalidation. Can we just wait for the user to close the buffer?

**Comment by naim94a** [comment]
I think this is done for now - Feedback would be appreciated at your convenience

File: crates/languages/src/c.rs [hunk 2]
```diff
@@ -292,6 +292,40 @@ impl super::LspAdapter for CLspAdapter {
...
+    ) {
+        if let Some((server_id, buffer)) = buffer_access {
+            let snapshot = buffer.snapshot();
+            let inactive_regions = buffer
+                .get_diagnostics(server_id)
+                .into_iter()
+                .flat_map(|v| v.iter())
+                .filter(|diag| clangd_ext::is_inactive_region(&diag.diagnostic))
+                .map(move |diag| {
+                    // try to reconstruct inactiveRegion diagnostic messages...
```
Comment (SomeoneToIgnore) [inline crates/languages/src/c.rs:309]:
```suggestion
```

File: crates/project/src/lsp_store.rs [hunk 3]
```diff
@@ -449,8 +448,18 @@ impl LocalLspStore {
                 move |mut params, cx| {
                     let adapter = adapter.clone();
                     if let Some(this) = this.upgrade() {
-                        adapter.process_diagnostics(&mut params);
                         this.update(cx, |this, cx| {
+                            {
+                                let buffer = params
+                                    .uri
+                                    .to_file_path()
```
Comment (SomeoneToIgnore) [inline crates/project/src/lsp_store.rs:455]: This is a bit trickier than just this, let's use

https://github.com/zed-industries/zed/blob/d253d46fdf894b520db5d19fc3e4ba7c5ba439f4/crates/project/src/lsp_command.rs#L1213-L1218 

instead?

File: crates/languages/src/c.rs [hunk 4]
```diff
@@ -292,6 +292,40 @@ impl super::LspAdapter for CLspAdapter {
         }
         Ok(original)
     }
+
+    fn process_diagnostics(
+        &self,
+        params: &mut lsp::PublishDiagnosticsParams,
+        buffer_access: Option<(LanguageServerId, &'_ Buffer)>,
```
Comment (SomeoneToIgnore) [inline crates/languages/src/c.rs:299]:
```suggestion
        server_id: LanguageServerId,
        buffer: Option<&'_ Buffer>,
```

`LanguageServerId` is not about buffers, it's about that `lsp::PublishDiagnosticsParams` above and should always be present, as the message emitter's id.

File: crates/project/src/lsp_store.rs [hunk 5]
```diff
@@ -5886,6 +5895,18 @@ impl LspStore {
         version: Option<i32>,
         diagnostics: Vec<DiagnosticEntry<Unclipped<PointUtf16>>>,
         cx: &mut Context<Self>,
+    ) -> Result<(), anyhow::Error> {
```
Comment (SomeoneToIgnore) [inline crates/project/src/lsp_store.rs:5898]:
```suggestion
    ) -> anyhow::Result<()> {
```

Or just `Result<()>` as most of the code around imports this type.

File: crates/languages/src/rust.rs [hunk 6]
```diff
@@ -945,7 +949,7 @@ mod tests {
                 },
             ],
         };
-        RustLspAdapter.process_diagnostics(&mut params);
+        RustLspAdapter.process_diagnostics(&mut params, None);
```
Comment (SomeoneToIgnore) [inline crates/languages/src/rust.rs:952]: We could add a test too now, when we have logic added in `c.rs`.

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you, I like the idea to provide buffer state access in the buffer-related server responses instead of another static to mutate.

Left a few comments, seems close to be merged.

File: crates/project/src/lsp_store.rs [hunk 7]
```diff
@@ -5900,6 +5921,29 @@ impl LspStore {
...
+                    .get_diagnostics(server_id)
+                    .into_iter()
+                    .flat_map(|diag| {
+                        diag.iter().filter(|v| filter(&v.diagnostic)).map(|v| {
+                            let start = Unclipped(v.range.start.to_point_utf16(&snapshot));
+                            let end = Unclipped(v.range.end.to_point_utf16(&snapshot));
+                            let range = start..end;
+                            DiagnosticEntry {
+                                range,
+                                diagnostic: v.diagnostic.clone(),
```
[the remark this answers is not in the document]
Comment (naim94a) [reply crates/project/src/lsp_store.rs:5939]: `DiagnosticSet::iter()` only iterates on references, so we cannot move the `diagnostic` elements.

File: crates/project/src/lsp_store.rs (hunk 3 above)
Comment (naim94a) [reply crates/project/src/lsp_store.rs:455]: I'm not sure how. `open_local_buffer_via_lsp` returns a future.
I tried:
- `cx.background_executor().block(this. open_local_buffer_via_lsp(...))`
- creating a one-shot `channel()`, and calling `cx.spawn(...)`

but both caused Zed to freeze up.

Comment (SomeoneToIgnore) [reply crates/project/src/lsp_store.rs:455]: Oh, odd that we use `cx: &mut AsyncApp` here but the closure is synchronous.
Even more odd, `on_request` returns `impl Future`.

I think we should change this and use the method from above, for a "safe" way to get the buffer.
Right now, seems that we are not in the wrong though, as the action is expected to be run from the editor with this buffer, so it should be preloaded (on a client too) and can be accessed by path.

------------

Just for the reference: we could've used cx.spawn/background_spawn/spawn_in to get a `Task<T>` out that runs until dropped.
`.detach()` would consume it and run it to completion at some point, storing it as a field allows "run last to the end" patterns, etc.

In this case, storing such tasks looks bothersome and `.detach()` — uncertain for me state-wise, esp. given the first paragraph.

**Review by SomeoneToIgnore — APPROVED** [review]
Thank you a lot for persistence.
